File: FNM_RE/stages/diagnostics.py
"""FNM_RE 第五阶段：diagnostic 投影。"""

from __future__ import annotations

from typing import Any

from FNM_RE.models import (
    ChapterRecord,
    DiagnosticEntryRecord,
    DiagnosticNoteRecord,
    DiagnosticPageRecord,
    NoteRegionRecord,
    Phase4Structure,
    TranslationUnitRecord,
    UnitPageSegmentRecord,
    UnitParagraphRecord,
)
from FNM_RE.shared.refs import extract_note_refs, replace_frozen_refs

_ERROR_TRANSLATION_STATUSES = {"error", "retry_pending", "retrying", "manual_required"}
_DONE_UNIT_STATUSES = {"done", "done_manual"}


def _raw_print_page_label(page_no: int, pages: list[dict]) -> str:
    for page in pages or []:
        if int(page.get("bookPage") or 0) != int(page_no):
            continue
        label = str(page.get("printPageLabel") or "").strip()
        if label:
            return label
        print_page = page.get("printPage")
        try:
            parsed = int(print_page)
        except (TypeError, ValueError):
            parsed = 0
        if parsed > 0:
            return str(parsed)
        break
    return str(page_no)


def _display_pages_label(page_no: int, pages: list[dict]) -> str:
    raw = _raw_print_page_label(page_no, pages)
    if not raw:
        return ""
    return raw if raw.startswith("原书 p.") else f"原书 p.{raw}"


def _segment_from_any(payload: UnitPageSegmentRecord | dict[str, Any]) -> UnitPageSegmentRecord:
    if isinstance(payload, UnitPageSegmentRecord):
        return payload
    paragraphs: list[UnitParagraphRecord] = []
    for row in list((payload or {}).get("paragraphs") or []):
        if isinstance(row, UnitParagraphRecord):
            paragraphs.append(row)
            continue
        paragraphs.append(
            UnitParagraphRecord(
                order=int((row or {}).get("order", len(paragraphs) + 1) or len(paragraphs) + 1),
                kind=str((row or {}).get("kind") or ""),
                heading_level=int((row or {}).get("heading_level", 0) or 0),
                source_text=str((row or {}).get("source_text") or ""),
                display_text=str((row or {}).get("display_text") or (row or {}).get("source_text") or ""),
                cross_page=(row or {}).get("cross_page"),
                consumed_by_prev=bool((row or {}).get("consumed_by_prev")),
                section_path=list((row or {}).get("section_path") or []),
                print_page_label=str((row or {}).get("print_page_label") or ""),
                translated_text=str((row or {}).get("translated_text") or ""),
                translation_status=str((row or {}).get("translation_status") or "pending"),
                attempt_count=max(0, int((row or {}).get("attempt_count", 0) or 0)),
                last_error=str((row or {}).get("last_error") or ""),
                manual_resolved=bool((row or {}).get("manual_resolved")),
            )
        )
    return UnitPageSegmentRecord(
        page_no=int((payload or {}).get("page_no") or 0),
        paragraph_count=int((payload or {}).get("paragraph_count") or 0),
        source_text=str((payload or {}).get("source_text") or ""),
        display_text=str((payload or {}).get("display_text") or (payload or {}).get("source_text") or ""),
        paragraphs=paragraphs,
    )


def _entry_status(paragraph: UnitParagraphRecord) -> str:
    status = str(paragraph.translation_status or "").strip()
    if status in _ERROR_TRANSLATION_STATUSES:
        return "error"
    if str(paragraph.translated_text or "").strip():
        return "done"
    return "pending"


def _build_diagnostic_entry(
    *,
    page_no: int,
    paragraph: UnitParagraphRecord,
    pages: list[dict],
) -> DiagnosticEntryRecord:
    source_text = str(paragraph.display_text or paragraph.source_text or "").strip()
    translated_text = str(paragraph.translated_text or "").strip()
    translation_status = str(paragraph.translation_status or "").strip() or ("done" if translated_text else "pending")
    entry_text = replace_frozen_refs(translated_text or source_text)
    refs: list[dict[str, str]] = []
    seen_refs: set[tuple[str, str]] = set()
    for candidate in (
        paragraph.source_text,
        paragraph.display_text,
        paragraph.translated_text,
        source_text,
        translated_text,
    ):
        for ref in extract_note_refs(str(candidate or "").strip()):
            key = (str(ref.get("kind") or ""), str(ref.get("note_id") or ""))
            if key in seen_refs:
                continue
            seen_refs.add(key)
            refs.append({"kind": key[0], "note_id": key[1]})
    return DiagnosticEntryRecord(
        original=replace_frozen_refs(source_text),
        translation=entry_text,
        footnotes="",
        footnotes_translation="",
        heading_level=int(paragraph.heading_level or 0),
        pages=_display_pages_label(page_no, pages),
        _startBP=int(page_no),
        _endBP=int(page_no),
        _printPageLabel=_raw_print_page_label(page_no, pages),
        _status=_entry_status(paragraph),
        _error=str(paragraph.last_error or ""),
        _translation_source="manual"
        if bool(paragraph.manual_resolved)
        else ("model" if translated_text else "source"),
        _machine_translation=entry_text if translated_text else "",
        _manual_translation=entry_text if bool(paragraph.manual_resolved) else "",
        _cross_page=paragraph.cross_page,
        _section_path=list(paragraph.section_path or []),
        _fnm_refs=refs,
        _note_kind="",
        _note_marker="",
        _note_number=None,
        _note_section_title="",
        _note_confidence=0.0,
        _translation_status=translation_status,
        _attempt_count=max(0, int(paragraph.attempt_count or 0)),
        _manual_resolved=bool(paragraph.manual_resolved),
    )


def _note_unit_by_note_id(translation_units: list[TranslationUnitRecord]) -> dict[str, TranslationUnitRecord]:
    payload: dict[str, TranslationUnitRecord] = {}
    for unit in translation_units:
        if str(unit.kind or "") not in {"footnote", "endnote"}:
            continue
        note_id = str(unit.note_id or "").strip()
        if note_id:
            payload[note_id] = unit
    return payload


def _chapter_meta_by_id(phase4: Phase4Structure) -> dict[str, ChapterRecord]:
    return {
        str(chapter.chapter_id or "").strip(): chapter
        for chapter in phase4.chapters
        if str(chapter.chapter_id or "").strip()
    }


def _note_region_by_id(phase4: Phase4Structure) -> dict[str, NoteRegionRecord]:
    return {
        str(region.region_id or "").strip(): region
        for region in phase4.note_regions
        if str(region.region_id or "").strip()
    }


def build_diagnostic_projection(
    phase4: Phase4Structure,
    translation_units: list[TranslationUnitRecord],
    *,
    pages: list[dict],
    only_pages: list[int] | None = None,
) -> tuple[list[DiagnosticPageRecord], list[DiagnosticNoteRecord], dict[str, Any]]:
    visible_page_filter = {int(page_no) for page_no in (only_pages or []) if int(page_no) > 0}
    page_rows: dict[int, DiagnosticPageRecord] = {}

    for unit in translation_units:
        if str(unit.kind or "") != "body":
            continue
        source_meta = {
            "section_id": unit.section_id,
            "section_title": unit.section_title,
            "section_start_page": unit.section_start_page,
            "section_end_page": unit.section_end_page,
            "unit_id": unit.unit_id,
        }
        for raw_segment in unit.page_segments:
            segment = _segment_from_any(raw_segment)
            page_no = int(segment.page_no or 0)
            if page_no <= 0:
                continue
            if visible_page_filter and page_no not in visible_page_filter:
                continue
            page_row = page_rows.setdefault(
                page_no,
                DiagnosticPageRecord(
                    _pageBP=page_no,
                    _status="pending",
                    pages=_display_pages_label(page_no, pages),
                    _page_entries=[],
                    _fnm_source=dict(source_meta),
                ),
            )
            for paragraph in segment.paragraphs:
                if paragraph.consumed_by_prev:
                    continue
                page_row._page_entries.append(
                    _build_diagnostic_entry(
                        page_no=page_no,
                        paragraph=paragraph,
                        pages=pages,
                    )
                )
            if any(entry._status == "error" for entry in page_row._page_entries):
                page_row._status = "error"
            elif any(entry._status == "done" for entry in page_row._page_entries):
                page_row._status = "done"
            else:
                page_row._status = "pending"

    diagnostic_pages = [page_rows[page_no] for page_no in sorted(page_rows)]

    chapter_by_id = _chapter_meta_by_id(phase4)
    note_region_by_id = _note_region_by_id(phase4)
    note_units = _note_unit_by_note_id(translation_units)
    diagnostic_notes: list[DiagnosticNoteRecord] = []
    for item in phase4.note_items:
        note_id = str(item.note_item_id or "").strip()
        if not note_id:
            continue
        chapter_id = str(item.chapter_id or "").strip()
        chapter = chapter_by_id.get(chapter_id)
        region = note_region_by_id.get(str(item.region_id or "").strip())
        note_kind = str(region.note_kind if region else "")
        unit = note_units.get(note_id)
        if not note_kind:
            unit_kind = str(unit.kind if unit else "").strip().lower()
            if unit_kind in {"footnote", "endnote"}:
                note_kind = unit_kind
            else:
                note_kind = "endnote" if str(note_id).lower().startswith("en-") else "footnote"
        start_page = int(item.page_no or (unit.page_start if unit else 0) or 0)
        diagnostic_notes.append(
            DiagnosticNoteRecord(
                note_id=note_id,
                section_id=chapter_id,
                section_title=str((chapter.title if chapter else chapter_id) or chapter_id),
                section_start_page=int((chapter.start_page if chapter else 0) or 0),
                section_end_page=int((chapter.end_page if chapter else 0) or 0),
                kind=note_kind,
                original_marker=str(item.marker or ""),
                start_page=start_page,
                pages=[start_page] if start_page > 0 else [],
                source_text=str(item.text or (unit.source_text if unit else "") or ""),
                translated_text=str((unit.translated_text if unit else "") or ""),
                translate_status=str((unit.status if unit else "pending") or "pending"),
                region_id=str(item.region_id or ""),
            )
        )
    diagnostic_notes.sort(
        key=lambda row: (
            int(row.section_start_page or 0),
            int(row.start_page or 0),
            str(row.kind or ""),
            str(row.note_id or ""),
        )
    )

    page_entry_count = sum(len(page._page_entries) for page in diagnostic_pages)
    translated_page_count = sum(
        1
        for page in diagnostic_pages
        if any(entry._translation_source != "source" for entry in page._page_entries)
    )
    diagnostic_summary = {
        "diagnostic_page_summary": {
            "page_count": len(diagnostic_pages),
            "entry_count": page_entry_count,
            "error_page_count": sum(1 for page in diagnostic_pages if page._status == "error"),
            "translated_page_count": translated_page_count,
        },
        "diagnostic_note_summary": {
            "note_count": len(diagnostic_notes),
            "translated_count": sum(
                1
                for note in diagnostic_notes
                if str(note.translate_status or "") in _DONE_UNIT_STATUSES
                or bool(str(note.translated_text or "").strip())
            ),
            "pending_count": sum(
                1
                for note in diagnostic_notes
                if str(note.translate_status or "") not in _DONE_UNIT_STATUSES
                and str(note.translate_status or "") not in _ERROR_TRANSLATION_STATUSES
                and not bool(str(note.translated_text or "").strip())
            ),
            "error_count": sum(
                1
                for note in diagnostic_notes
                if str(note.translate_status or "") in _ERROR_TRANSLATION_STATUSES
            ),
        },
    }
    return diagnostic_pages, diagnostic_notes, diagnostic_summary

File: fnm_re_rs/fnm-phase6/src/diagnostics.rs
//! Diagnostic 投影。
//!
//! ←→ Python `FNM_RE/stages/diagnostics.py` (336 行 / 10 函数)

use std::collections::HashMap;

use fnm_core::records::{
    ChapterRecord, DiagnosticEntryRecord, DiagnosticNoteRecord, DiagnosticPageRecord,
    NoteRegionRecord, Phase4Structure, TranslationUnitRecord, UnitPageSegmentRecord,
    UnitParagraphRecord,
};
use fnm_core::refs::{extract_note_refs, replace_frozen_refs, EndnoteMode};

/// 错误翻译状态集合
static _ERROR_TRANSLATION_STATUSES: &[&str] =
    &["error", "retry_pending", "retrying", "manual_required"];

/// 已完成翻译状态集合
static _DONE_UNIT_STATUSES: &[&str] = &["done", "done_manual"];

/// 从 pages 提取 {bookPage: printPageLabel} 轻量映射。
///
/// ←→ Python `build_print_page_map()` (diagnostics.py:24)
pub fn build_print_page_map(pages: &[serde_json::Value]) -> HashMap<i64, String> {
    let mut result = HashMap::new();
    for page in pages {
        let bp = page.get("bookPage").and_then(|v| v.as_i64()).unwrap_or(0);
        if bp <= 0 {
            continue;
        }
        let label = page
            .get("printPageLabel")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .trim()
            .to_string();
        let label = if label.is_empty() {
            page.get("printPage")
                .and_then(|v| v.as_i64())
                .filter(|&pp| pp > 0)
                .map(|pp| pp.to_string())
                .unwrap_or_else(|| bp.to_string())
        } else {
            label
        };
        result.insert(bp, label);
    }
    result
}

/// 获取原始打印页标签。
///
/// ←→ Python `_raw_print_page_label()` (diagnostics.py:42)
pub fn raw_print_page_label(page_no: i64, print_pages: &HashMap<i64, String>) -> String {
    print_pages
        .get(&page_no)
        .cloned()
        .unwrap_or_else(|| page_no.to_string())
}

/// 获取显示用的页码标签。
///
/// ←→ Python `_display_pages_label()` (diagnostics.py:63)
pub fn display_pages_label(page_no: i64, print_pages: &HashMap<i64, String>) -> String {
    let raw = raw_print_page_label(page_no, print_pages);
    if raw.is_empty() {
        return String::new();
    }
    if raw.starts_with("原书 p.") {
        raw
    } else {
        format!("原书 p.{}", raw)
    }
}

/// 从任意格式转换为 UnitPageSegmentRecord。
///
/// ←→ Python `_segment_from_any()` (diagnostics.py:70)
pub fn segment_from_any(payload: &serde_json::Value) -> UnitPageSegmentRecord {
    // 如果已经是 UnitPageSegmentRecord 类型（通过 serde_json 转换）
    if let Ok(segment) = serde_json::from_value::<UnitPageSegmentRecord>(payload.clone()) {
        return segment;
    }

    // 否则手动构建
    let paragraphs: Vec<UnitParagraphRecord> = payload
        .get("paragraphs")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .enumerate()
                .map(|(idx, row)| {
                    if let Ok(p) = serde_json::from_value::<UnitParagraphRecord>(row.clone()) {
                        return p;
                    }
                    UnitParagraphRecord {
                        order: row
                            .get("order")
                            .and_then(|v| v.as_i64())
                            .unwrap_or(idx as i64 + 1),
                        kind: row
                            .get("kind")
                            .and_then(|v| v.as_str())
                            .unwrap_or("")
                            .to_string(),
                        heading_level: row
                            .get("heading_level")
                            .and_then(|v| v.as_i64())
                            .unwrap_or(0),
                        source_text: row
                            .get("source_text")
                            .and_then(|v| v.as_str())
                            .unwrap_or("")
                            .to_string(),
                        display_text: row
                            .get("display_text")
                            .and_then(|v| v.as_str())
                            .or_else(|| row.get("source_text").and_then(|v| v.as_str()))
                            .unwrap_or("")
                            .to_string(),
                        cross_page: row
                            .get("cross_page")
                            .cloned()
                            .unwrap_or(serde_json::Value::Null),
                        consumed_by_prev: row
                            .get("consumed_by_prev")
                            .and_then(|v| v.as_bool())
                            .unwrap_or(false),
                        section_path: row
                            .get("section_path")
                            .and_then(|v| v.as_array())
                            .map(|arr| {
                                arr.iter()
                                    .filter_map(|v| v.as_str().map(|s| s.to_string()))
                                    .collect()
                            })
                            .unwrap_or_default(),
                        print_page_label: row
                            .get("print_page_label")
                            .and_then(|v| v.as_str())
                            .unwrap_or("")
                            .to_string(),
                        translated_text: row
                            .get("translated_text")
                            .and_then(|v| v.as_str())
                            .unwrap_or("")
                            .to_string(),
                        translation_status: row
                            .get("translation_status")
                            .and_then(|v| v.as_str())
                            .unwrap_or("pending")
                            .to_string(),
                        attempt_count: row
                            .get("attempt_count")
                            .and_then(|v| v.as_i64())
                            .unwrap_or(0)
                            .max(0),
                        last_error: row
                            .get("last_error")
                            .and_then(|v| v.as_str())
                            .unwrap_or("")
                            .to_string(),
                        manual_resolved: row
                            .get("manual_resolved")
                            .and_then(|v| v.as_bool())
                            .unwrap_or(false),
                    }
                })
                .collect()
        })
        .unwrap_or_default();

    UnitPageSegmentRecord {
        page_no: payload.get("page_no").and_then(|v| v.as_i64()).unwrap_or(0),
        paragraph_count: payload
            .get("paragraph_count")
            .and_then(|v| v.as_i64())
            .unwrap_or(0),
        source_text: payload
            .get("source_text")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        display_text: payload
            .get("display_text")
            .and_then(|v| v.as_str())
            .or_else(|| payload.get("source_text").and_then(|v| v.as_str()))
            .unwrap_or("")
            .to_string(),
        paragraphs,
    }
}

/// 获取段落的翻译状态。
///
/// ←→ Python `_entry_status()` (diagnostics.py:105)
pub fn entry_status(paragraph: &UnitParagraphRecord) -> String {
    let status = paragraph.translation_status.trim();
    if _ERROR_TRANSLATION_STATUSES.contains(&status) {
        return "error".to_string();
    }
    if !paragraph.translated_text.trim().is_empty() {
        return "done".to_string();
    }
    "pending".to_string()
}

/// 构建诊断条目。
///
/// ←→ Python `_build_diagnostic_entry()` (diagnostics.py:114)
pub fn build_diagnostic_entry(
    page_no: i64,
    paragraph: &UnitParagraphRecord,
    print_pages: &HashMap<i64, String>,
) -> DiagnosticEntryRecord {
    let source_text = {
        let display = paragraph.display_text.trim();
        if !display.is_empty() {
            display.to_string()
        } else {
            paragraph.source_text.trim().to_string()
        }
    };
    let translated_text = paragraph.translated_text.trim().to_string();
    let translation_status = if !paragraph.translation_status.trim().is_empty() {
        paragraph.translation_status.trim().to_string()
    } else if !translated_text.is_empty() {
        "done".to_string()
    } else {
        "pending".to_string()
    };

    let entry_text = if !translated_text.is_empty() {
        replace_frozen_refs(&translated_text, EndnoteMode::Legacy)
    } else {
        replace_frozen_refs(&source_text, EndnoteMode::Legacy)
    };

    let mut refs: Vec<serde_json::Value> = Vec::new();
    let mut seen_refs: std::collections::HashSet<(String, String)> =
        std::collections::HashSet::new();

    for candidate in &[
        &paragraph.source_text,
        &paragraph.display_text,
        &paragraph.translated_text,
        &source_text,
        &translated_text,
    ] {
        for ref_item in extract_note_refs(candidate) {
            let key = (ref_item.kind.as_str().to_string(), ref_item.note_id.clone());
            if seen_refs.contains(&key) {
                continue;
            }
            seen_refs.insert(key.clone());
            refs.push(serde_json::json!({
                "kind": key.0,
                "note_id": key.1,
            }));
        }
    }

    DiagnosticEntryRecord {
        original: replace_frozen_refs(&source_text, EndnoteMode::Legacy),
        translation: entry_text.clone(),
        footnotes: String::new(),
        footnotes_translation: String::new(),
        heading_level: paragraph.heading_level,
        pages: display_pages_label(page_no, print_pages),
        _start_bp: page_no,
        _end_bp: page_no,
        _print_page_label: raw_print_page_label(page_no, print_pages),
        _status: entry_status(paragraph),
        _error: paragraph.last_error.clone(),
        _translation_source: if paragraph.manual_resolved {
            "manual".to_string()
        } else if !translated_text.is_empty() {
            "model".to_string()
        } else {
            "source".to_string()
        },
        _machine_translation: if !translated_text.is_empty() {
            entry_text.clone()
        } else {
            String::new()
        },
        _manual_translation: if paragraph.manual_resolved {
            entry_text.clone()
        } else {
            String::new()
        },
        _cross_page: paragraph.cross_page.clone(),
        _section_path: paragraph.section_path.clone(),
        _fnm_refs: refs,
        _note_kind: String::new(),
        _note_marker: String::new(),
        _note_number: None,
        _note_section_title: String::new(),
        _note_confidence: 0.0,
        _translation_status: translation_status,
        _attempt_count: paragraph.attempt_count.max(0),
        _manual_resolved: paragraph.manual_resolved,
    }
}

/// 构建 note_id → TranslationUnitRecord 映射。
///
/// ←→ Python `_note_unit_by_note_id()` (diagnostics.py:170)
pub fn note_unit_by_note_id(
    translation_units: &[TranslationUnitRecord],
) -> HashMap<String, &TranslationUnitRecord> {
    let mut payload = HashMap::new();
    for unit in translation_units {
        if unit.kind != "footnote" && unit.kind != "endnote" {
            continue;
        }
        let note_id = unit.note_id.trim().to_string();
        if !note_id.is_empty() {
            payload.insert(note_id, unit);
        }
    }
    payload
}

/// 构建 chapter_id → ChapterRecord 映射。
///
/// ←→ Python `_chapter_meta_by_id()` (diagnostics.py:181)
pub fn chapter_meta_by_id(phase4: &Phase4Structure) -> HashMap<String, &ChapterRecord> {
    phase4
        .chapters
        .iter()
        .filter(|ch| !ch.chapter_id.trim().is_empty())
        .map(|ch| (ch.chapter_id.trim().to_string(), ch))
        .collect()
}

/// 构建 region_id → NoteRegionRecord 映射。
///
/// ←→ Python `_note_region_by_id()` (diagnostics.py:189)
pub fn note_region_by_id(phase4: &Phase4Structure) -> HashMap<String, &NoteRegionRecord> {
    phase4
        .note_regions
        .iter()
        .filter(|r| !r.region_id.trim().is_empty())
        .map(|r| (r.region_id.trim().to_string(), r))
        .collect()
}

/// 构建诊断投影。
///
/// ←→ Python `build_diagnostic_projection()` (diagnostics.py:197)
pub fn build_diagnostic_projection(
    phase4: &Phase4Structure,
    translation_units: &[TranslationUnitRecord],
    print_page_map: Option<&HashMap<i64, String>>,
    only_pages: Option<&[i64]>,
) -> (
    Vec<DiagnosticPageRecord>,
    Vec<DiagnosticNoteRecord>,
    serde_json::Value,
) {
    let empty_map = HashMap::new();
    let print_pages = print_page_map.unwrap_or(&empty_map);

    let visible_page_filter: std::collections::HashSet<i64> = only_pages
        .map(|pages| pages.iter().filter(|&&p| p > 0).copied().collect())
        .unwrap_or_default();

    let mut page_rows: HashMap<i64, DiagnosticPageRecord> = HashMap::new();

    // 处理 body units
    for unit in translation_units {
        if unit.kind != "body" {
            continue;
        }
        let source_meta = serde_json::json!({
            "section_id": unit.section_id,
            "section_title": unit.section_title,
            "section_start_page": unit.section_start_page,
            "section_end_page": unit.section_end_page,
            "unit_id": unit.unit_id,
        });

        for raw_segment in &unit.page_segments {
            let segment = if let Ok(s) = serde_json::from_value::<UnitPageSegmentRecord>(
                serde_json::to_value(raw_segment).unwrap_or_default(),
            ) {
                s
            } else {
                segment_from_any(&serde_json::to_value(raw_segment).unwrap_or_default())
            };

            let page_no = segment.page_no;
            if page_no <= 0 {
                continue;
            }
            if !visible_page_filter.is_empty() && !visible_page_filter.contains(&page_no) {
                continue;
            }

            let page_row = page_rows
                .entry(page_no)
                .or_insert_with(|| DiagnosticPageRecord {
                    _page_bp: page_no,
                    _status: "pending".to_string(),
                    pages: display_pages_label(page_no, print_pages),
                    _page_entries: Vec::new(),
                    _fnm_source: source_meta.clone(),
                });

            for paragraph in &segment.paragraphs {
                if paragraph.consumed_by_prev {
                    continue;
                }
                page_row._page_entries.push(build_diagnostic_entry(
                    page_no,
                    paragraph,
                    print_pages,
                ));
            }

            if page_row._page_entries.iter().any(|e| e._status == "error") {
                page_row._status = "error".to_string();
            } else if page_row._page_entries.iter().any(|e| e._status == "done") {
                page_row._status = "done".to_string();
            }
        }
    }

    let mut diagnostic_pages: Vec<DiagnosticPageRecord> = page_rows.into_values().collect();
    diagnostic_pages.sort_by_key(|p| p._page_bp);

    // 处理 notes
    let chapter_by_id = chapter_meta_by_id(phase4);
    let note_region_by_id_map = note_region_by_id(phase4);
    let note_units = note_unit_by_note_id(translation_units);

    let mut diagnostic_notes: Vec<DiagnosticNoteRecord> = Vec::new();
    for item in &phase4.note_items {
        let note_id = item.note_item_id.trim().to_string();
        if note_id.is_empty() {
            continue;
        }
        let chapter_id = item.chapter_id.trim().to_string();
        let chapter = chapter_by_id.get(&chapter_id);
        let region = note_region_by_id_map.get(item.region_id.trim());
        let unit = note_units.get(&note_id);

        let note_kind = region
            .map(|r| r.note_kind.as_str().to_string())
            .unwrap_or_else(|| item.note_kind.as_str().to_string());

        let start_page = item.page_no.max(unit.map(|u| u.page_start).unwrap_or(0));

        diagnostic_notes.push(DiagnosticNoteRecord {
            note_id,
            section_id: chapter_id.clone(),
            section_title: chapter
                .map(|ch| ch.title.clone())
                .unwrap_or_else(|| chapter_id),
            section_start_page: chapter.map(|ch| ch.start_page).unwrap_or(0),
            section_end_page: chapter.map(|ch| ch.end_page).unwrap_or(0),
            kind: note_kind,
            original_marker: item.marker.clone(),
            start_page,
            pages: if start_page > 0 {
                vec![start_page]
            } else {
                vec![]
            },
            source_text: {
                let text = item.text.trim();
                if !text.is_empty() {
                    text.to_string()
                } else {
                    unit.map(|u| u.source_text.clone()).unwrap_or_default()
                }
            },
            translated_text: unit.map(|u| u.translated_text.clone()).unwrap_or_default(),
            translate_status: unit
                .map(|u| u.status.clone())
                .unwrap_or_else(|| "pending".to_string()),
            region_id: item.region_id.clone(),
        });
    }

    diagnostic_notes.sort_by(|a, b| {
        a.section_start_page
            .cmp(&b.section_start_page)
            .then(a.start_page.cmp(&b.start_page))
            .then(a.kind.cmp(&b.kind))
            .then(a.note_id.cmp(&b.note_id))
    });

    // 统计
    let page_entry_count: usize = diagnostic_pages.iter().map(|p| p._page_entries.len()).sum();
    let translated_page_count = diagnostic_pages
        .iter()
        .filter(|p| {
            p._page_entries
                .iter()
                .any(|e| e._translation_source != "source")
        })
        .count();

    let note_count = diagnostic_notes.len();
    let translated_count = diagnostic_notes
        .iter()
        .filter(|n| {
            _DONE_UNIT_STATUSES.contains(&n.translate_status.as_str())
                || !n.translated_text.trim().is_empty()
        })
        .count();
    let pending_count = diagnostic_notes
        .iter()
        .filter(|n| {
            !_DONE_UNIT_STATUSES.contains(&n.translate_status.as_str())
                && !_ERROR_TRANSLATION_STATUSES.contains(&n.translate_status.as_str())
                && n.translated_text.trim().is_empty()
        })
        .count();
    let error_count = diagnostic_notes
        .iter()
        .filter(|n| _ERROR_TRANSLATION_STATUSES.contains(&n.translate_status.as_str()))
        .count();

    let diagnostic_summary = serde_json::json!({
        "diagnostic_page_summary": {
            "page_count": diagnostic_pages.len(),
            "entry_count": page_entry_count,
            "error_page_count": diagnostic_pages.iter().filter(|p| p._status == "error").count(),
            "translated_page_count": translated_page_count,
        },
        "diagnostic_note_summary": {
            "note_count": note_count,
            "translated_count": translated_count,
            "pending_count": pending_count,
            "error_count": error_count,
        },
    });

    (diagnostic_pages, diagnostic_notes, diagnostic_summary)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_build_print_page_map() {
        let pages = vec![
            serde_json::json!({"bookPage": 1, "printPageLabel": "i"}),
            serde_json::json!({"bookPage": 2, "printPage": 2}),
            serde_json::json!({"bookPage": 0}), // 应该被跳过
        ];
        let result = build_print_page_map(&pages);
        assert_eq!(result.len(), 2);
        assert_eq!(result.get(&1).unwrap(), "i");
        assert_eq!(result.get(&2).unwrap(), "2");
    }

    #[test]
    fn test_raw_print_page_label() {
        let mut pages = HashMap::new();
        pages.insert(1, "i".to_string());
        assert_eq!(raw_print_page_label(1, &pages), "i");
        assert_eq!(raw_print_page_label(99, &pages), "99");
    }

    #[test]
    fn test_display_pages_label() {
        let mut pages = HashMap::new();
        pages.insert(1, "i".to_string());
        assert_eq!(display_pages_label(1, &pages), "原书 p.i");
        assert_eq!(display_pages_label(99, &pages), "原书 p.99");
    }

    #[test]
    fn test_entry_status() {
        let mut p = UnitParagraphRecord::default();
        p.translation_status = "error".to_string();
        assert_eq!(entry_status(&p), "error");

        p.translation_status = "done".to_string();
        p.translated_text = "translated".to_string();
        assert_eq!(entry_status(&p), "done");

        p.translation_status = "pending".to_string();
        p.translated_text = "".to_string();
        assert_eq!(entry_status(&p), "pending");
    }

    #[test]
    fn test_note_unit_by_note_id() {
        let units = vec![
            TranslationUnitRecord {
                kind: "body".to_string(),
                note_id: "n1".to_string(),
                ..Default::default()
            },
            TranslationUnitRecord {
                kind: "footnote".to_string(),
                note_id: "n2".to_string(),
                ..Default::default()
            },
        ];
        let result = note_unit_by_note_id(&units);
        assert_eq!(result.len(), 1);
        assert!(result.contains_key("n2"));
    }

    #[test]
    fn test_chapter_meta_by_id() {
        let phase4 = Phase4Structure {
            chapters: vec![ChapterRecord {
                chapter_id: "ch1".to_string(),
                title: "Chapter 1".to_string(),
                start_page: 0,
                end_page: 0,
                pages: vec![],
                source: fnm_core::types::ChapterSource::Fallback,
                boundary_state: fnm_core::types::BoundaryState::Ready,
            }],
            ..Default::default()
        };
        let result = chapter_meta_by_id(&phase4);
        assert_eq!(result.len(), 1);
        assert!(result.contains_key("ch1"));
    }
}
